runners/direct-java/src/main/java/org/apache/beam/runners/direct/ExecutorServiceParallelExecutor.java

```diff
@@ -364,7 +366,12 @@ public void run() {
           } else if (update.getException().isPresent()) {
             visibleUpdates.offer(VisibleExecutorUpdate.fromThrowable(update.getException().get()));
           }
-          update = allUpdates.poll();
+          if (numUpdates >= MAX_UPDATES_PER_CALL) {
```

Comment:
Updates are expected to take an approximately constant amount of time? Otherwise, using time might be better.